substrate-node/pallets/pallet-smart-contract/src/tests.rs

```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+        Balances::transfer_allow_death(RuntimeOrigin::signed(bob()), charlie(), bob_balance).unwrap();
+
+        // The grace period for the rent contract should start
+        // Node contract should also transition to grace period
+        pool_state
+            .write()
+            .should_call_bill_contract(rent_contract_id, Ok(Pays::Yes.into()), 21);
+        run_to_block(21, Some(&mut pool_state));
+
+        // Check if evenets contains ContractGracePeriodStarted events for both contracts
+        let events = System::events();
+        log::debug!("Events: {:?}", events);
+        assert_eq!(
+            events
+                .iter()
+                .filter(|e| matches!(
+                    e.event,
+                    MockEvent::SmartContractModule(SmartContractEvent::ContractGracePeriodStarted { .. })
+                ))
+                .count(),
+            2
+        );
+        pool_state
+            .write()
+            .should_call_bill_contract(node_contract_id, Ok(Pays::Yes.into()), 22);
+        run_to_block(22, Some(&mut pool_state));
+
+        // Advance 10 cycles
+        // Typically, the rent contract is canceled immediately after the grace period expires
+        // However, if one of the associated node contracts hasn't yet distributed rewards,
+        // the rent contract should remain active until the node contract completes billing and reward distribution
+        // Once done, the rent contract can then be canceled
+        // Case 3: see https://github.com/threefoldtech/tfchain/issues/1002
+        for i in 1..=10 {
+            let block_number = 21 + i * BillingFrequency::get();
+            pool_state.write().should_call_bill_contract(
+                rent_contract_id,
+                Ok(Pays::Yes.into()),
+                block_number,
+            );
+            run_to_block(block_number, Some(&mut pool_state));
+            pool_state.write().should_call_bill_contract(
+                node_contract_id,
+                Ok(Pays::Yes.into()),
+                block_number+1,
+            );
+            run_to_block(block_number+1, Some(&mut pool_state));
+        }
+        // Check if the events contains RewardDistributed event
+        let events = System::events();
+        log::debug!("Events: {:?}", events);
+        assert_eq!(
+            events
+                .iter()
+                .filter(|e| matches!(
+                    e.event,
+                    MockEvent::SmartContractModule(SmartContractEvent::RewardDistributed { .. })
+                ))
+                .count(),
+            1
+        );
+        // Check if the last event is NodeContractCanceled event
+        assert_eq!(
+            events
+                .iter()
+                .filter(|e| matches!(
+                    e.event,
+                    MockEvent::SmartContractModule(SmartContractEvent::NodeContractCanceled { .. })
+                ))
+                .count(),
+            1
+        );
```

Comment:
Checking for last event could be done like this:

~~~rust
        assert_eq!(
            events
                .last()
                .unwrap()
                .event,
            MockEvent::SmartContractModule(SmartContractEvent::{
                    contract_id: node_contract_id,
                    node_id,
                    twin_id: bob_twin_id
                })
        );
~~~